app/controllers/courses_controller.rb

```diff
@@ -59,9 +59,21 @@ def update
 
   def destroy
     validate
-    DeleteCourseWorker.schedule_deletion(course: @course, current_user:)
-    render json: { success: true }
-  end
+    if params.key?(:campaign)
```

Comment:
Can the flow control be simplified here? I don't like having two separate `if` blocks within the method.

Reply:
```
  def destroy
    validate
    if params.key?(:campaign)
      remove_course_from_campaign
    else
      DeleteCourseWorker.schedule_deletion(course: @course, current_user: current_user)
      render json: { success: true }
    end
  end
  
  def remove_course_from_campaign
    campaigns_course = CampaignsCourses.find_by(course_id: @course.id, campaign_id: params[:campaign_id])
    result = campaigns_course.destroy
    message = result ? 'campaign.course_removed' : 'campaign.course_already_removed'
    flash[:notice] = t(message, title: @course.title, campaign_title: params[:campaign_title])
    DeleteCourseWorker.schedule_deletion(course: @course, current_user: current_user)
    redirect_to programs_campaign_path(params[:slug])
  end
```

made a new function `remove_course_from_campaign` to **increase readability** and **simplified the workflow**

Reply:
This flow is better, but there is duplication of the `schedule_deletion` part and the remove_course_from_campaign method does more than just remove the campaign, it also queues the course for deletion.

I think both versions of the delete action should be moved to a new child controller.